Contract implementation and tests

nockty opened the pull request:
Contracts are defined by teams at the beginning of the game.

Commit:
:construction: Add contract implementation

Commit:
:white_check_mark: Add test helper

Commit:
:white_check_mark: Add tests for contracts

Commit:
:green_heart: Add state and move interfaces

Scrabouillmax commented on server/src/main/java/com/coincoinche/engine/Contract.java:
```diff
@@ -0,0 +1,177 @@
+  private Player player;
+
+  private Contract(int value, boolean capot, boolean generale, Suit suit) {
+    this.value = value;
```
Can we add a tiny comment here to precise that value is -1 in case of 'capot' or 'generale'?

Scrabouillmax commented on server/src/main/java/com/coincoinche/engine/Contract.java:
```diff
@@ -0,0 +1,177 @@
+   * @return true if the other contract is strictly lower.
+   */
+  public boolean isHigherThan(Contract other) {
+    // handle case where other is null
```
```suggestion
```
Can delete that :)

Scrabouillmax commented on server/src/main/java/com/coincoinche/engine/Contract.java:
```diff
@@ -0,0 +1,177 @@
+  private Suit suit;
+  private Player player;
+
+  private Contract(int value, boolean capot, boolean generale, Suit suit) {
```
Would it be worth creating an enum ContractValue, same as for suit?
Otherwise, can we at least create a static method `isValid(value)` to check if a value is valid for a contract?
We can be explicit in the constructor then

nockty replied:
Agreed

Scrabouillmax commented on server/src/test/java/com/coincoinche/engine/ContractTest.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class ContractTest {
+
+  private class TestCase extends CommonTestHelper.TestCase {
```
Can we use clearer naming here? 
```suggestion
  private class ContractEqualsTestCase extends CommonTestHelper.TestCase {
```

I love this helper :p 
Can we add a test for the isHigherThan method as well in this file? It should be really fast with the helper:

```suggestion
  private class ContractEqualsTestCase extends CommonTestHelper.TestCase {
 (...)
 private class ContractIsHigherThanTestCase extends CommonTestHelper.TestCase {
```

nockty replied:
I'll just move these classes to the methods directly

Scrabouillmax commented on server/src/test/java/com/coincoinche/engine/CommonTestHelper.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Several helper methods for initializing the game.
+ */
+public class CommonTestHelper {
```
Clearer naming?
```suggestion
public class GameSetupTestHelper {
```

nockty replied:
It's not used only to set up the game though. I will change the javadoc comment which is outdated.

Scrabouillmax reviewed (commented):
I love the test helper 🥇 

Just a few comments mostly to clarify namings ;)

Commit:
:recycle: Refactor contracts

Use an enum to encode the value and include capot and generale
with other values by the way.

Commit:
:white_check_mark: Update contract tests

Add a test for isHigherThan.

Commit:
:speech_balloon: Update Contract's toString

Scrabouillmax reviewed (approved):
🚀
